### PR TITLE
Active Storage eager loading: support more methods

### PR DESCRIPTION
**Pull request by ghiculescu**
As noted at https://github.com/rails/rails/pull/40842#issuecomment-895231378, the current implementation works great with the `processed` method, but doesn't work with other methods such as `key`. This PR fixes that, by adding the missing `includes` statements.

cc @collimarco

### CODE REVIEW

**Comment by collimarco** [comment]
Wow, **thank you so much!**

I was struggling to find a solution... I couldn't find where in the call chain the blob was loaded again. Now from your work I guess that the `record.image` association had no idea about the `blob` association and thus loaded it again, while if we access to the `blob` directly from the `VariantWithRecord`, we still have access to it. 

I have quickly tested this in our application and everything seems to work, perfect.

**Comment by ghiculescu** [comment]
Hmmm, this has broken some tests. Don't ship it just yet 😆

File: activestorage/test/controllers/representations/redirect_controller_test.rb [hunk 1]
```diff
@@ -60,8 +60,7 @@ class ActiveStorage::Representations::RedirectControllerWithVariantsWithStrictLo
     follow_redirect!
     assert_match(/^inline/, response.headers["Content-Disposition"])
 
-    @blob.reload # became free of strict_loading?
-    image = read_image(@blob.variant(resize_to_limit: [100, 100]))
+    image = read_image_from_response
```
Comment (ghiculescu) [inline activestorage/test/controllers/representations/redirect_controller_test.rb:63]: note to self: `ActiveStorage::Representations::RedirectController`

`@representation.record.blob` doesn't match the variant blob?

**Comment by collimarco** [comment]
There's one thing that I don't understand in the new code: (let `@transformed=false`) calling variant `key` is like calling `blob.key`... Isn't that the key of the original file? In this case we need the key of the variant, not the key of the original uploaded file.

**Comment by collimarco** [comment]
As in my previous comment, I confirm that we need to call `record&.image.key` and not `record&.blob.key`, because they are different (one is the key of the variant, the other the key of the original file).

The `record&.image.class` is `ActiveStorage::Attached::One`... however I can't find the definition of `#key` in that class. Any idea where it is defined? (I have even tried `record&.image.method(:key).source_location`, but it is empty)

**Comment by collimarco** [comment]
@georgeclaghorn I see that you originally wrote the code for `VariantWithRecord`. Any hint on how to generate the variant key without hitting the database again? (considering that the associations Attachment -> Blog -> Variants are already eager loaded using `includes`)

**Comment by ghiculescu** [comment]
I'll have some more time tomorrow to work on this

**Comment by ghiculescu** [comment]
Simplified my approach, the solution is just to add the missing `includes`.

**Comment by collimarco** [comment]
Thanks for this improvement!

So in order to get the `item.image.variant({}).key` we need to:

1. Load the Attachment
2. Load the Blob (original file)
3. Load the Variant
4. Load another Attachment (image of the Variant)
5. Load another Blob (the Variant image file)

That's really deep nesting... I expected to be able to compute the Variant key directly from the Variant, I didn't expected that the Variant doesn't represent the file, but it needs to be attached again to the actual file (Blob).

I don't understand this choice, but I guess there are good reasons for that. Also this is not a decision made now by this pull request, but it was made when the Variant was stored in the database.

In any case I have tested this PR in production (together with https://github.com/rails/rails/pull/40842) and it seems to work great! Despite the deep nesting, I've seen a 10x improvement on pages that have hundreds of images. Thanks again!

**Comment by collimarco** [comment]
Can someone merge this? It's really useful for performance

**Comment by ghiculescu** [comment]
@pixeltrix could I bother you for a review here? It's just fixing a bug in https://github.com/rails/rails/pull/40842

**Comment by daveallie** [comment]
@ghiculescu while you're here, can you please apply the same fixes to the `has_one_attached` version of ActiveStorage attachments: https://github.com/rails/rails/blob/b7eb4bdfea1128340b701d79dc5190d193d645be/activestorage/lib/active_storage/attached/model.rb#L73

They suffer from the same N+1 issue when doing bulk queries, and would benefit from the same deep `includes` call.

**Comment by ghiculescu** [comment]
@rafaelfranca this fixes a bug that was found in 7.0 alpha, could it please be included in the next RC?

**Comment by collimarco** [comment]
Keep it open. Please someone should merge this...

**Review by jorgemanrubia — APPROVED** [review]
Good one @ghiculescu 👍. We'll get this one merged (cc @jeremy).

**Comment by ghiculescu** [comment]
Thanks!

**Comment by collimarco** [comment]
Great, thanks 🎉

**Pull request merged**